app/src/main/java/com/example/dispatcher/domain/repository/article/ArticlesRepositoryImpl.kt

```diff
@@ -0,0 +1,43 @@
+package com.example.dispatcher.domain.repository.article
+
+import android.content.Context
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import com.example.dispatcher.data.api.ApiConfigManager
+import com.example.dispatcher.data.api.Secrets
+import com.example.dispatcher.data.api.news.NewsServiceApi
+import com.example.dispatcher.data.model.news.TopHeadlines
+import com.example.dispatcher.data.network.NetworkManager
+import retrofit2.Call
+import retrofit2.Callback
+import retrofit2.Response
+
+class ArticlesRepositoryImpl(context: Context) : IArticleRepository {
+
+    private val newsServiceApi = NetworkManager.createService(NewsServiceApi::class.java)
+
+    private val topHeadlinesLiveData = MutableLiveData<TopHeadlines?>()
+
+    override fun fetchArticles(): LiveData<TopHeadlines?> {
+        newsServiceApi.getTopHeadlines(ApiConfigManager.getCountryCode(), Secrets.API_KEY).enqueue(object : Callback<TopHeadlines> {
+            override fun onResponse(call: Call<TopHeadlines>, response: Response<TopHeadlines>) {
+                if (response.isSuccessful) {
+                    response.body()?.let { topHeadlines ->
+
+                        topHeadlinesLiveData.postValue(topHeadlines)
+                    } ?: run {
+                        topHeadlinesLiveData.postValue(null)
```

Comment:
in case of an error, it's more convenient to write the failure logic only in the onFailure function and then inside the onResponse you'll call onFailure, so the logic is in one place.